### PR TITLE
PLANNER-2569 Improve performance of disabled constraints 

### PR DESCRIPTION
**Pull request by triceo**
Depends on https://github.com/kiegroup/optaplanner/pull/1646

### CODE REVIEW

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/stream/DroolsConstraintStreamScoreDirectorFactory.java [hunk 1]
```diff
@@ -180,36 +185,22 @@ public boolean isDroolsAlphaNetworkCompilationEnabled() {
...
-            String constraintId = (String) Objects.requireNonNull(
-                    rule.getMetaData().get(CONSTRAINT_ID_RULE_METADATA_KEY),
-                    () -> "Impossible state: Rule ("
-                            + ConstraintMatchTotal.composeConstraintId(rule.getPackageName(), rule.getName())
-                            + ") has no constraint ID.");
-            return !disabledConstraintIdSet.contains(constraintId);
+            return ruleEnabledCache.computeIfAbsent(rule, r -> {
+                String constraintId = ConstraintMatchTotal.composeConstraintId(r.getPackageName(), r.getName());
+                return !disabledConstraintIdSet.contains(constraintId);
+            });
```
Comment (triceo) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/stream/DroolsConstraintStreamScoreDirectorFactory.java:203]: Reading 2 collections becomes reading just one collection, plus 1 write during cold start.
This should speed up this code considerably.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/stream/DroolsConstraintStreamScoreDirectorFactory.java [hunk 2]
```diff
@@ -151,15 +150,21 @@ private static KieBase buildKieBaseFromModel(Model model, boolean droolsAlphaNet
...
+                kieSession.setGlobal(globalName, scoreInliner.buildWeightedScoreImpacter(constraint));
             }
-            String globalName = kieBaseDescriptor.getConstraintToGlobalMap().get(constraint).getName();
-            kieSession.setGlobal(globalName, scoreInliner.buildWeightedScoreImpacter(constraint));
-        });
+        }
         // Return only the inliner as that holds the work product of the individual impacters.
-        return new SessionDescriptor<>(kieSession, new ConstraintDisablingAgendaFilter(constraintToWeightMap), scoreInliner);
+        AgendaFilter agendaFilter =
+                disabledConstraints.isEmpty() ? null : new ConstraintDisablingAgendaFilter(Set.copyOf(disabledConstraints));
```
Comment (triceo) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/stream/DroolsConstraintStreamScoreDirectorFactory.java:166]: The agenda filter execution will be skipped entirely when there are no rules to disable.

**Comment by rsynek** [comment]
As SonarCloud suggests, this field:
https://github.com/kiegroup/optaplanner/blob/da168d0c64a909208a9dc3db9d1c8b14b6e22bbc/optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/stream/DroolsConstraintStreamScoreDirectorFactory.java#L68
seems to be unused.

**Review by rsynek — APPROVED** [review]
LGTM

**Comment by triceo** [comment]
jenkins retest this

**Pull request merged**